clash-prelude/tests/Clash/Tests/Fixed.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+module Clash.Tests.Fixed (tests) where
+
+import qualified Prelude as P
+import Test.Tasty
+import Test.Tasty.QuickCheck hiding (Fixed)
+
+import Clash.Prelude hiding (span)
+
+-- Saturate a number to be within certain bounds according to SaturationMode.
+saturate
+  :: RealFrac a
+  => a
+  -- minBound
+  -> a
+  -- maxBound
+  -> a
+  -- Span: the repetition interval of wrapping behaviour.
+  -> SaturationMode
+  -> a -> a
+saturate minB _ span SatWrap x =
+  let offs = x - minB
+      offsW = minB + (offs - span * fromIntegral (floor @_ @Integer
+                (offs / span)))
+  in offsW
+saturate minB maxB _ SatBound x
+  | x < minB  = minB
+  | x > maxB  = maxB
+  | otherwise = x
+saturate minB maxB _ SatZero x
+  | x < minB  = 0
+  | x > maxB  = 0
+  | otherwise = x
+saturate minB maxB _ SatSymmetric x
+  | x < minB  = if minB < 0 then (-maxB) else minB
+  | x > maxB  = maxB
+  | otherwise = x
+
+-- Saturate to bounds of type b.
+--
+-- It assumes the following values for the span (the range of numbers in the
+-- bounded type):
+--   - 2*(-minBound) when minBound < 0
+--   - floor maxBound + 1 otherwise
+-- This is correct for Unsigned, Signed and Fixed.
+saturateToBounded
+  :: forall b
+   . (Bounded b, Real b)
+  => SaturationMode
+  -> Rational
+  -> Rational
+saturateToBounded w x =
+  let span = case (minBound @b) < 0 of
+               True  -> 2 * negate (toRational (minBound @b))
+               False -> 1 + toRational (floor @Double @Integer $ fromRational
+                              $ toRational $ maxBound @b)
+  in saturate (toRational (minBound @b)) (toRational (maxBound @b)) span w x
+
+-- Verify satSucc for a list of values.
+satSuccPropList
+  :: forall a
+   . (Real a, SaturatingNum a)
+  => SaturationMode -> [a] -> Property
+satSuccPropList w es
+  = P.map (toRational . satSucc w) es
+    === P.map (saturateToBounded @a w . (+ 1) . toRational) es
+
+-- A list of Fixed numbers close to the bounds of the type.
+--
+-- Warning: this enumerates all fractional values, so it should only be used
+-- with types Fixed rep int frac where frac is reasonably small!
+closeToBounds
+  :: forall f rep int frac
+   . ( NumFixedC rep int frac
+     , f ~ Fixed rep int frac)
+  => [ f ]
```

Comment:
@martijnbastiaan , should I rewrite my tests to use `Hedgehog.Gen` instead of enumerating all interesting cases?

Reply:
Yes, if would have the same effect for small types, while making it possible to test larger ones too :).